### PR TITLE
Improve html parsing

### LINKED ISSUE #15: Improve HTML Parsing
### Description

Currently we just strip down the HTML received from Discourse (except  tags) and send the telegram message with no formatting. We also directly put the image links into the message body, and since they are usually quite long links they take valuable character space and look unpleasant. Discourse sends emojis as images also, this causes the bot to convert all the emojis into links and the telegram's link preview feature results in unwanted behavior.

To solve these issues, we can use the telegram's limited html tag support. Most of the tags can be mapped to the tags which telegram supports (e.g. \<h1\> to \<b\>). Telegram also support <a> tags, these can be used for links (including images). Telegram supports emojis natively so we just need to replace the entire emoji tag with appropriate unicode bytes.

#### Things To Do
- [ ] Formatting with Telegram-supported tags
- [ ] Link Handling
- [ ] Emoji Handling

### Related Files
[newsparser.py](https://github.com/kadircet/COWNotifier/blob/bc560d98a63c83b68f4a856b0101f3bc2858d5c4/newsparser.py)

### PR DESCRIPTION
**Pull request by hbostan**
Modified `articleParser` to add formatting to Telegram messages. It also shrinks down the image links by wrapping the name and size of the image in <a> tags. More over it replaces emojis in the html with their unicode values.

Added `emoji_codepoints.py`. It contains the byte values for emoji names

Added `HTMLSplitter`. It splits html into chunks with max size 4096. Generated chunks are also valid as html. This is needed to keep the formatting consistent across messages longer than 4096 characters.

### CODE REVIEW

File: newsparser.py [hunk 1]
```diff
@@ -29,25 +30,69 @@ def getHumanReadableDate(date):
...
+  #               (<a><img></a>) In this case parser removes the <img> tag
+  #               completely since Telegram handles <a> tags pretty well with
+  #               previews. (Also telegram doesn't support <img> tags)
+  # Handle Emoji: Emojis are represented as <img> tags but they are not wrapped
+  #               in <a> tags, unlike the previous case. Parser replaces <img>
+  #               tags with 'class="emoji"' with their unicode codepoint.
+
+  # Tags supported by Telegram
+  supported_tags = [
+      'strike', 'i', 'strong', 'b', 'pre', 'ins', 'a', 'code', 'del', 's', 'em',
```
Comment (kadircet) [inline newsparser.py:54]: would be nice if you could put a link to some telegram bot api page

File: newsparser.py [hunk 2]
```diff
@@ -29,25 +30,69 @@ def getHumanReadableDate(date):
...
 
 class articleParser(HTMLParser):
+  # articleParser parses the raw html from Discourse and generates html that
+  # Telegram likes. During the process it performs the following.
+  #
+  # Convert Tags: Telegram supports a very small subset of html tags. In order
+  #               to provide (some) formatting in the telegram message, parser
+  #               converts tags that telegram doesn't like into ones it does
+  #               or removes them (not the contents) if no conversion exists.
+  # Handle Image: Images are wrapped in <a> tags with href same as img's src.
```
Comment (kadircet) [inline newsparser.py:40]: that's also nice, but it limits the message to have only a single image and lots of links. i think there's value in showing multiple images, for example as attachments.

no need to do it in this PR though, could you just put a todo and open an issue?

File: newsparser.py [hunk 3]
```diff
@@ -29,25 +30,69 @@ def getHumanReadableDate(date):
...
     self.reset()
     self.fed = []
 
+  def handle_entityref(self, name):
+    # This is done so that html entities like '&gt;' remain escaped.
+    self.fed.append(f'&{name};')
+
   def handle_starttag(self, tag, attrs):
-    if tag != "img":
+    tag_attrs = {at[0]: at[1] for at in attrs if at[0] in self.useful_attrs}
```
Comment (kadircet) [inline newsparser.py]: why not just create a mapping for all attributes, dropping the need for `useful_attrs`

File: newsparser.py [hunk 4]
```diff
@@ -29,25 +30,69 @@ def getHumanReadableDate(date):
...
+    ptag = tag
```
Comment (kadircet) [inline newsparser.py]: can we do something like:
```
if tag == 'a':
  href = tag_attrs.get('href', '')
  if not href:
     log the tag and attr, as I suppose this should not happen
  self.append('<a href="{}">'.format(href)) # or whatever format string magic is there in python
  self.open_tags.append('a')
  return
if tag == 'img':
   check if it is an emoji, bail out if not
   also if it is not an emoji, check that last open_tag is 'a', and log the tag if not
   return
if tag not in supported_tags:
  tag = tag_mapping.get(tag)
if not tag:
  return
self.append(f'<{tag}>')
self.open_tags.append(tag)
```

File: newsparser.py [hunk 5]
```diff
@@ -29,25 +30,69 @@ def getHumanReadableDate(date):
...
-    self.fed.append(d)
+  def handle_data(self, data):
+    self.fed.append(data)
+
+  def handle_endtag(self, tag):
+    # If tag is not supported/converted we don't need it's closing tag since
+    # it is never opened.
+    if tag not in self.supported_tags:
+      tag = self.tag_mapping.get(tag, None)
+    if tag != None:
```
Comment (kadircet) [inline newsparser.py]:
```
if tag != self.open_tags[-1]:
  return
open_tags.pop
self.fed.append ...
```

File: html_splitter.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    return sum(len(f'<{tag+attrs}>') for tag, attrs in self.opened_tags)
+
+  def getClosingTagsLen(self):
+    return sum(len(f'</{tag}>') for tag, attrs in self.opened_tags)
+
+  def getCurrentMsgLen(self):
+    return sum(len(item) for item in self.current_msg)
+
+  def getTotalMsgLen(self):
+    return self.getClosingTagsLen() + self.getCurrentMsgLen()
```
Comment (kadircet) [inline html_splitter.py]: what about opening ?

File: emoji_codepoints.py [hunk 7]
```diff
@@ -1,4147 +1,8288 @@
 # -*- coding: UTF-8 -*-
```
Comment (kadircet) [inline emoji_codepoints.py]: how is this file generated ?

can we instead check-in the generator and run it on the server at startup, to populate a data file mentioned in the config, if it doesn't exist or empty

**Review by kadircet — CHANGES_REQUESTED** [review]

**Review by kadircet — APPROVED** [review]
As discussed offline let's get rid of the html splitter logic for now and keep the good old naive chunking.

File: newsparser.py (hunk 1 above)
Comment (hbostan) [reply newsparser.py:54]: Done

File: newsparser.py (hunk 2 above)
Comment (hbostan) [reply newsparser.py:40]: Added todo

File: newsparser.py (hunk 3 above)
Comment (hbostan) [reply newsparser.py]: done

File: newsparser.py (hunk 4 above)
Comment (hbostan) [reply newsparser.py]: done

File: newsparser.py (hunk 5 above)
Comment (hbostan) [reply newsparser.py]: done

**Pull request merged**